### PR TITLE
Optimize Audit Policy Actions

### PR DESCRIPTION
**Pull request by yamaceay**
# AuditPolicy action parsing & optimization

## Summary

Introduce a robust audit-action parsing and optimization layer, extend the AuditPolicy client and reconciler to support richer principal-aware reads and multi-query CREATEs, and add new end-to-end tests and fixtures for AuditPolicy, Role, Usergroup and related resources.

This PR normalizes, groups, and validates audit action strings so the controller and client produce predictable SQL and behavior when working with principals and complex action clauses.

## What changed

- New audit action parsing/optimization utilities:
  - `OptimizeAuditActions` and helpers to parse action strings (supports `FOR PRINCIPALS`, `EXCEPT FOR`, `USER`, `USERGROUP`).
  - Internal types: `parsedAction`, `parsedPrincipal` and stringify/grouping helpers.
- `internal/clients/hana/auditpolicy/auditpolicy_client.go`:
  - Read now scans principal columns (`PRINCIPAL_NAME`, `EXCEPT_PRINCIPAL_NAME`, `PRINCIPAL_TYPE`) and reconstructs full action strings.
  - `prepareCreateSql` now returns multiple SQL statements (one per action variant); `Create` runs each query.
  - `getSelectSql` updated to select principal-related columns.
- `internal/controller/auditpolicy/reconciler.go`:
  - Parameter builders now return errors.
  - Reconciler calls `OptimizeAuditActions` for observed/desired inputs, and bubbles errors with logging.
- Tests & fixtures:

## Behavioral notes

- `OptimizeAuditActions` normalizes and groups actions, but currently returns an error if grouping yields more than one grouped action (this is an intentional guard).
- CREATE logic will issue multiple `CREATE AUDIT POLICY` variants when multiple action strings are present.
- Read paths now include principal context, so observed `AuditActions` contain `FOR PRINCIPALS` / `EXCEPT FOR PRINCIPALS` fragments when applicable.

## How to validate

Manual smoke test:
- Apply the CR: `kubectl apply -f {path/to/auditpolicy.yaml}`
- Observe the provider logs and the created DB audit policy; check `AuditActions` in the CR status reflect normalized `FOR/EXCEPT` principal fragments.

## Migration & rollout notes

- Existing policies with complex multi-action combinations may hit the current grouping guard and return an error — review policies that expect multiple independent grouped actions.
- If you depend on the old, un-normalized string formatting, validate consumer integrations that read `AuditActions` from the CR status.

## Risks & follow-ups

- Relax grouping limit: support multiple grouped actions instead of erroring.
- Add unit tests for parsing edge cases (user vs usergroup, comma spacing, capitalization, unexpected tokens).
- Add SQL escaping tests for `PolicyName` and principal values.

### CODE REVIEW

**Review by baronzhangdev — COMMENTED** [review]
If you agree, I’d be happy to help fix and further improve this feature in this PR.

I also noticed the PR is currently missing unit tests, and I can contribute a set of UTs as part of my commits.

**Review by baronzhangdev — APPROVED** [review]
LGTM

File: internal/controller/auditpolicy/reconciler.go [hunk 1]
```diff
@@ -316,20 +334,29 @@ func (c *external) Delete(ctx context.Context, mg resource.Managed) (managed.Ext
...
 }
 
-func buildDesiredParameters(cr *v1alpha1.AuditPolicy) *v1alpha1.AuditPolicyParameters {
+func buildDesiredParameters(cr *v1alpha1.AuditPolicy) (*v1alpha1.AuditPolicyParameters, error) {
+	auditActions, err := auditpolicy.OptimizeAuditActions(cr.Spec.ForProvider.AuditActions)
+	if err != nil {
+		return nil, err
+	}
 	return &v1alpha1.AuditPolicyParameters{
 		PolicyName:          strings.ToUpper(cr.Spec.ForProvider.PolicyName),
```
Comment (baronzhangdev) [inline internal/controller/auditpolicy/reconciler.go]: I noticed that `PolicyName` in the AuditPolicy controller also uses `strings.ToUpper()` conversion. According to the [HANA documentation](https://help.sap.com/docs/SAP_HANA_PLATFORM/4fe29514fd584807ac9f2a04f6754767/20d3d56075191014af43d6487fcaa603.html?locale=en-US\&state=PRODUCTION\&version=2.0.08\&q=auditpolicy#example), policy names should preserve their original case when using double-quoted identifiers.

I've addressed a similar issue for Role and DbSchema in #74. We might want to remove the forced uppercase conversion for `PolicyName` as well and add corresponding unit tests. Would you be able to take a look at this?

**Review by baronzhangdev — APPROVED** [review]
Thanks. LGTM.
